docs/design_notes/lift.md

```diff
@@ -0,0 +1,355 @@
+For example: `vmap` has broadcasted and vectorized arguments, while `scan` has scan, carry, and broadcast arguments.
+A variable would have to define its behavior for all these transformations otherwise a `Module` would not be compatible with
+these transformations. Alternatively, we would have to make default decisions for how transformations are handled.
+However, this could lead to silent bugs because the behavior might not actually be valid given the users intent. 
+
+The lift package also provides a general purpose `transform`, which allows an arbitrary function to transform a variable collection.
+For example, this can be used to tie the weights in a tied auto-encoder by transposing the weights.
+It is unclear whether a similar general purpose transform could be defined if lifting decisions were made at variable definition.
+
+
+### Linen
+
+The lifting module does not know about the Linen `Module` API.
+Instead it operates directly on instances of `flax.core.Scope`.
+A `Scope` instance contains the variables and PRNG sequences of a `Module`.
+Each `Module` instance has a `Scope` instance in the `.scope` field if it has a parent or it was created using `init` or `apply`.
+Typically, the top-level `Module` instance on which you call `init` or `apply` is the only `Module` instance that does not have a `Scope` bound to it.
+
+When a `Module` is transformed, we use the `flax.core.lift` APIs to lift the scope and use `Module.clone()` to create a new `Module` instance with the lifted scope bound to it.
+
+`flax.linen.transforms` exposes wrappers for the transformations in `flax.core.lift`. The core lifting APIs operate on functions while
+the Linen wrappers can transform either a `Module` class or a `Module` method.
+
+Thus, lifting is implemented independently from the Linen API. This separation of concern simplifies the implementation, while potentially allowing alternative `Module` abstractions to build upon a common core for lifting and state management.
+
+
+### Implementation
+
+The `pack(fn, in_vars, out_vars, rngs)` API goes through the following stages:
+
+
+1. *Scope de-duplication*
+
+    This stage is only relevant if multiple Scopes are lifted together.
+    In this case we must first find the set of root scopes.
+    A scope is a root if none of its ancestors are in the set of scopes that need to be lifted.
+
+    By only lifting roots we avoid lifting the same variables twice.
+
+    For non-root scopes we store a reference to its ancestor scope and a path such that we can later reconstruct it (stage 4).
+
+2. *Filter stage*
+
+    Variables and PRNG sequences are split up into groups. This way `fn` can lift each group into the transformation separately.
+    A group is defined by a filter specified as:
+    - a list of collections/prng names
+    - `True` (match everything)
+    - `False` (match nothing)
+    - `DenyList(filter)` (match everything but the specified filter).
```

Comment:
Do you meant to say `DenyList(list_of_collections)` instead of `DenyList(filter)`?

Reply:
DenyList can compose with an arbitrary filter (we need this internally).

Reply:
The only compositions that are possible are: DenyList(list), DenyList(True), DenyList(False), DenyList(DenyList(...)), right? Of these only the first one seems to make sense. Also, since the name was `DenyList` and `DenyFilter`, I assume you meant to use it with a list.

Reply:
Added some clarification